### PR TITLE
Refactor AddressingMode

### PR DESCRIPTION
**Pull request by Valgard**

### CODE REVIEW

File: src/emulator/instruction.ts [hunk 1]
```diff
@@ -213,91 +213,195 @@ function apply(opcode: number, instruction: Partial<Instruction>): void {
...
-apply(0x31, { operation: Operation.ld, addressingMode: AddressingMode.reg16_imm16, par1: r16.sp, cycles: 3, len: 3 });
-apply(0x21, { operation: Operation.ld, addressingMode: AddressingMode.reg16_imm16, par1: r16.hl, cycles: 3, len: 3 });
-apply(0x11, { operation: Operation.ld, addressingMode: AddressingMode.reg16_imm16, par1: r16.de, cycles: 3, len: 3 });
-apply(0x01, { operation: Operation.ld, addressingMode: AddressingMode.reg16_imm16, par1: r16.bc, cycles: 3, len: 3 });
-apply(0x36, { operation: Operation.ld, addressingMode: AddressingMode.ind8_imm8, par1: r16.hl, cycles: 3, len: 2 });
-apply(0xea, { operation: Operation.ld, addressingMode: AddressingMode.immind8_reg8, par2: r8.a, cycles: 4, len: 3 });
-apply(0xf0, { operation: Operation.ld, addressingMode: AddressingMode.reg8_imm8io, par1: r8.a, cycles: 3, len: 2 });
-apply(0xe0, { operation: Operation.ld, addressingMode: AddressingMode.imm8io_reg8, par2: r8.a, cycles: 3, len: 2 });
-apply(0xe2, { operation: Operation.ld, addressingMode: AddressingMode.reg8io_reg8, par1: r8.c, cycles: 2, len: 1 });
-apply(0xf2, { operation: Operation.ld, addressingMode: AddressingMode.reg8_reg8io, par2: r8.c, cycles: 2, len: 1 });
```
Comment (Valgard) [inline src/emulator/instruction.ts:257]: here was an bug: missing par2 leads always to F and not to A

File: src/emulator/cpu.ts [hunk 2]
```diff
@@ -336,119 +336,130 @@ export class Cpu {
...
-                break;
+        return this.getArg(instruction.par1);
+    }
 
-            case AddressingMode.ind8_reg8:
-            case AddressingMode.ind8_imm8:
-            case AddressingMode.ind8:
-                this.bus.write(this.state.r16[instruction.par1], value & 0xff);
-                break;
+    private getArg2(instruction: Instruction): number {
```
Comment (DirtyHairy) [inline src/emulator/cpu.ts:398]: I would just get rid of this as it is a mere trampoline with an overhead that is negligible in the best case. The null check will be obsolete with my other changes suggested below.

File: src/emulator/cpu.ts [hunk 3]
```diff
@@ -336,119 +336,130 @@ export class Cpu {
...
-        switch (instruction.addressingMode) {
-            case AddressingMode.reg8:
-            case AddressingMode.reg8_imm8:
-            case AddressingMode.reg8_ind8:
-            case AddressingMode.reg8_reg8:
-            case AddressingMode.reg8_imm8io:
-            case AddressingMode.reg8_reg8io:
-                this.state.r8[instruction.par1] = value & 0xff;
-                break;
+    private getArg1(instruction: Instruction): number {
```
Comment (DirtyHairy) [inline src/emulator/cpu.ts:390]: I would just get rid of this as it is a mere trampoline with an overhead that is negligible in the best case. The null check will be obsolete with my other changes suggested below.

File: src/emulator/cpu.ts [hunk 4]
```diff
@@ -336,119 +336,130 @@ export class Cpu {
...
-            case AddressingMode.reg8_ind8:
-            case AddressingMode.reg8_reg8:
-            case AddressingMode.reg8_reg8io:
-            case AddressingMode.reg8_imm8io:
-                return this.state.r8[instruction.par1];
-
-            case AddressingMode.reg16_imm16:
-            case AddressingMode.reg16:
-                return this.state.r16[instruction.par1];
+                if (par.value === undefined) {
```
Comment (DirtyHairy) [inline src/emulator/cpu.ts:355]: This and the following null checks are a performance overhead that will be obsolete with other changes suggested below.

File: src/emulator/instruction.ts [hunk 5]
```diff
@@ -27,32 +27,27 @@ export const enum Operation {
...
-    imm8io_reg8,
-    reg8_imm8io,
-    reg8io_reg8,
-    reg8_reg8io,
+
+    imm16,
+    reg16,
+}
+
+export interface OperationParameter {
```
Comment (DirtyHairy) [inline src/emulator/instruction.ts:41]: I would like to make `value` non-optional and just preset it to 0 in order to keep this data structure as static as possible (it is used in the hot code path).

File: src/emulator/instruction.ts [hunk 6]
```diff
@@ -27,32 +27,27 @@ export const enum Operation {
...
+    addressingMode: AddressingMode;
 }
 
 export interface Instruction {
     opcode: number;
     operation: Operation;
-    addressingMode: AddressingMode;
-    par1: number;
-    par2: number;
+    par1?: OperationParameter;
```
Comment (DirtyHairy) [inline src/emulator/instruction.ts:49]: I would like to make `par1` and `par2` non-optional and just preset them to `implicit/0` in order to keep this data structure as static as possible (it is used in the hot code path).

File: src/emulator/instruction.ts [hunk 7]
```diff
@@ -27,32 +27,27 @@ export const enum Operation {
...
-    reg8io_reg8,
-    reg8_reg8io,
+
+    imm16,
+    reg16,
+}
+
+export interface OperationParameter {
+    value?: number;
+    addressingMode: AddressingMode;
```
Comment (DirtyHairy) [inline src/emulator/instruction.ts:43]: Maybe shorten this to`mode` in order to keep the individual definitions below in a single line.

File: src/emulator/instruction.ts [hunk 8]
```diff
@@ -27,32 +27,27 @@ export const enum Operation {
...
 }
 
 export interface Instruction {
     opcode: number;
     operation: Operation;
```
Comment (DirtyHairy) [inline src/emulator/instruction.ts:48]: Maybe shorten this to`op` in order to keep the individual definitions below in a single line.

File: src/emulator/instruction.ts [hunk 9]
```diff
@@ -213,91 +213,195 @@ function apply(opcode: number, instruction: Partial<Instruction>): void {
...
-        addressingMode: AddressingMode.implicit,
-        par1: 0,
-        par2: 0,
         cycles: 0,
         len: 1,
     };
 
 apply(0, { operation: Operation.nop, cycles: 1, len: 1 });
-apply(0xc3, { operation: Operation.jp, addressingMode: AddressingMode.imm16, cycles: 4, len: 3 });
+apply(0xc3, { operation: Operation.jp, par1: { addressingMode: AddressingMode.imm16 }, cycles: 4, len: 3 });
```
Comment (DirtyHairy) [inline src/emulator/instruction.ts:227]: I would like to shorten  a few fields (see above) in order to keep those definitions in single lines --- I find this pretty hard to read otherwise. Note that prettier will not reformat automatically to single lines, you need to remove the first break --- it cost me a bit of digging to discover this last time it started splitting them 😏

**Review by DirtyHairy — COMMENTED** [review]
Nice work! I have added a few comments. Imo we should also add a consistency check below the opcode definitions to make sure that no opcode tries to load both from memory --- this would break (due to the `p + 1`). This issue existed before, but there was no addressing mode that would have exposed it. Now, this constellation is at least possible in theory (and in case of bugs 😛 ).

File: src/emulator/cpu.ts (hunk 3 above)
Comment (DirtyHairy) [reply src/emulator/cpu.ts:390]: On second thought: let's keep it (without the check) --- keeps things readable and should be optimized out by the engine.

File: src/emulator/cpu.ts (hunk 2 above)
Comment (DirtyHairy) [reply src/emulator/cpu.ts:398]: On second thought: let's keep it (without the check) --- keeps things readable and should be optimized out by the engine.

File: src/emulator/instruction.ts (hunk 6 above)
Comment (DirtyHairy) [reply src/emulator/instruction.ts:49]: Another thing to think about: once `par1` and `par2` are not optional anymore the additional object could be removed and be reduced to plain fields `par1/2` and `mode1/2` again. This is less clean, but it would avoid an additional dereference on every instruction decode.

**Comment by Valgard** [comment]
I have a weird bug in the disassemble output.
grafik

**Comment by Valgard** [comment]
okay, found it quick, was my fault. i have accidentally set the len for invalid instructions to zero.
